execution_workflows/APAtrap/modules/de_apa.nf

```diff
@@ -0,0 +1,29 @@
+// Import generic module functions
+include { initOptions; saveFiles; getSoftwareName } from './functions'
+
+params.options = [:]
+def options    = initOptions(params.options)
+
+/*
+    Run the third step of APAtrap: deAPA to detect genes having significant changes
+    in APA site usage between conditions.
+*/
+process DE_APA {
+        publishDir "${params.outdir}/apatrap", mode: params.publish_dir_mode
+        container "docker.io/apaeval/apatrap:latest"
+
+        input:
+        path de_apa_input
+
+        output:
+        path "$de_apa_output", emit: ch_de_apa_output
+
+        script:
+        de_apa_output = "de_apa_output.txt"
+        """
+        #!/usr/bin/Rscript
+
+        library(deAPA)
+        deAPA('$de_apa_input', '$de_apa_output', 1, 1, 1, 1, 20)
+        """
```

Comment:
I don't think parameters should be hardcoded in the pipeline.  They should be defined and described in a config file (they can be left at default values for now though).

The 3rd & 4th arguments of the `deAPA` function are group1 & group2 respectively according to docs. Does setting both to 1 (i.e. comparing samples in group1 to samples in group1) produce valid comparisons *between* groups?

Reply:
@SamBryce-Smith  good point, thank you! Do you happen to know if we'll always provide two sample files for differential challenge? And if so, will these two sample files used for differential challenge also used for the other two challenges? Wondering since the samplesheet only has one pair of bam and bai files, but if we need two sample files for differential challenge then it seems like I'd have to change the samplesheet headers so that it allows for two bam and bai files so that the deAPA step would take in two samples  to produce valid comparisons between groups

Reply:
I'm not sure about whether it would be two samples for each group in the differential challenge. I would expect it to be more, but our workflows should be able to count how many samples per group from the sample sheet (if it is formatted correctly).

I'm pretty sure the samples used for the differential challenge will be different to those used for the other challenges though.

RE your sample sheet, I will make comment again on review later (about to finish for day today) but I think it could be structured more as a 'long' rather than 'wide' table. Actually how it is at the moment gives the incorrect relationships about the files:

samplesheet_example_files.csv as of 0c9e44b :

the files in bam1 & bam2 are actually two **distinct** samples (control & siRNA KD), but this table would suggest they are the same sample 'test'

| sample | bam1                            | bai1                                | bam2                          | bai2                              | bed                              |
|--------|---------------------------------|-------------------------------------|-------------------------------|-----------------------------------|----------------------------------|
| test   | path_to/siControl_R1_2genes.bam | path_to/siControl_R1_2genes.bam.bai | path_to/siSrsf3_R1_2genes.bam | path_to/siSrsf3_R1_2genes.bam.bai | path_to/vm18.mouse.genemodel.bed |


If this table was 'longer', you could keep the same template/columns and just add rows if you have > 2 samples. Then you could also easily annotate which group they belong to as well.

My suggestion would be something like below. I'd also remove the `bed` column and place it in a more general config (i.e. global parameter for this given run), as I think we'll want to process each sample with the same annotation.

| sample    | group   | bam                             | bai                                 |
|-----------|---------|---------------------------------|-------------------------------------|
| siCTL_1   | siCTL   | path_to/siControl_R1_2genes.bam | path_to/siControl_R1_2genes.bam.bai |
| siSrsf3_1 | siSrsf3 | path_to/siSrsf3_R1_2genes.bam   | path_to/siSrsf3_R1_2genes.bam.bai   |

Hopefully that makes sense, from personal experience I find it easier to deal with 'longer' tables anyway. I suspect you may have to rejig how you input info from sample table to pipeline but I think the extra flexibility is a bonus. Happy to provide suggestions/discuss if you feel you need it (though I'm out of my depth with nextflow) :)

P.S. - we have the test data in the main repository. If the example sample sheet could contain the relative paths to the test data already, that would make it that bit easier for us to do a test run of the pipeline!

Reply:
Agree with Sam, it would be awesome if the sample sheet could already point to the test data in the APAeval repo (Although I just tried it but couldn't make nextflow accept relative paths. Is that a reality?).    

Also, I agree that the genemodel file could be specified somewhere else (not in the sample sheet), as it is the same for all samples (and sample sheet should ideally only contain sample specific information).    

Speaking of the genemodel file, I'm not comfortable with downloading it from UCSC and adding "chr" to the bedgraph files, as this could make the outputs incomparable (well, not *really* incomparable, but it's still a bit dirty) to outputs from other tools that use ensembl annotations. We should generate the genemodel in pre-processing (best to add to utils, see #62 )

Reply:
I had the same problem with relative paths too. Adding the absolute path from `pwd` before the relative path worked for me though but felt a bit dirty (workflow was able to generate bedgraphs at least).

Reply:
Would make most sense to define the path in the config file, to not run into issues of path resolution rules.